feat(booking): implement cross-page/week navigation in booking details sheet with view persistence

eunjae-lee opened the pull request:
## Summary

follow-up of #25251

This PR enhances the bookings interface with three major improvements:

1. Cross-page/week navigation in booking details sheet: Users can now seamlessly navigate between bookings across different pages (list view) or weeks (calendar view) using next/previous buttons. The sheet automatically loads the next/previous period when reaching the end of the current data set.

2. UTM parameter tracking: The booking details sheet now displays UTM campaign parameters (utm_source, utm_medium, utm_campaign, utm_term, utm_content) when available.

3. Persistent view preference: The bookings interface now remembers the user's last selected view (list or calendar) using localStorage, maintaining their preference across sessions.

4. Feature flag check: since this PR, we check the feature flag on the user level, not on the team level (in preparation for user-level feature opt-in banner in an upcoming PR)

## Demo

### 1. Cross-page/week navigation in booking details sheet

https://github.com/user-attachments/assets/d7f08b1b-9f84-4dc6-8389-4e836cd24b32

### 2. UTM parameter tracking

<img width="524" height="974" alt="Screenshot 2025-12-04 at 14 37 12" src="https://github.com/user-attachments/assets/a97088f1-cd85-4041-b847-e55f3d24fe29" />

### 3. Persistent view preference

https://github.com/user-attachments/assets/12c811c2-59d2-48d5-879b-1160548d9998
















































<!-- This is an auto-generated description by cubic. -->
---
## Summary by cubic
Adds seamless next/previous navigation in the Booking Details sheet across pages and weeks with smart prefetching and nearest-week jumps. Also introduces a persistent list/calendar view with a WeekPicker and several booking detail/UI improvements, including auto-scroll and selection highlighting.

- **New Features**
  - Cross-page/week navigation in Booking Details Sheet with prefetch; sheet is deep-linkable via uid and store is pagination/week-aware. Buttons now disable when no bookings exist and jump directly to the week containing the nearest booking; auto-selects first/last item when crossing pages/weeks.
  - Calendar view uses infinite week loading, respects user week start, adds WeekPicker, status color bars, and shows timezone. Prefetches previous/next weeks; selected events are highlighted and auto-scrolled into view.
  - List view prefetches next page, uses separators and improved skeleton; selected bookings are highlighted and auto-scrolled into view. View toggle persists in localStorage and hides on mobile.
  - Booking details show payment/refund, assignment reason, and UTM/tracking; booking actions include Accept/Reject buttons and a Cancel dialog.

- **Refactors**
  - Adapter-based navigation with deferred selection to prevent duplicate TRPC calls; store adds capabilities, pending selection, and transition state.
  - Feature flag check moved to user-level (was team-level).
  - Use booking.uid (not id) across actions/tests and calendar events; add viewer.bookings.getBookingDetails; bookings.get supports statuses[], cursor, and sort; getAllUserBookings stays backward-compatible.
  - New containers/hooks for list and calendar; simplify filters (member-only on calendar for admins/owners); SSR-safe useMediaQuery moved to packages; UI polish across table, dropdowns, and badges.
  - DataTable tweaks: scroll to top on pagination and validate pageSize > 0; remove legacy view parser in favor of a hook.

<sup>Written for commit 51685102aca43421f352b9982ea9ae3bd7db64bc. Summary will update automatically on new commits.</sup>

<!-- End of auto-generated description by cubic. -->

eunjae-lee committed:
display UTM parameters

eunjae-lee committed:
persist view (list | calendar) in localStorage

eunjae-lee committed:
fix navigation on calendar view

eunjae-lee committed:
move to next page on clicking "next" of last item in the page

improve navigation

^ Conflicts:
^	apps/web/modules/bookings/components/BookingCalendarContainer.tsx

eunjae-lee committed:
Delete apps/web/modules/bookings/NAVIGATION_IMPLEMENTATION.md

eunjae-lee committed:
avoid page size being 0

eunjae-lee committed:
check feature flag on user level

CarinaWolli reviewed (commented):
When reaching the last booking in the list view the next button is disabled, which is good: 
<img width="131" height="58" alt="Screenshot 2025-12-09 at 12 52 54 PM" src="https://github.com/user-attachments/assets/97debf02-9611-46d5-b913-323700a85f9a" />

But in the calendar view, the button is not disabled and instead it closes by default. It should also be disabled when you reach the last booking

eunjae-lee committed:
use map to improve useBookingListData

eunjae-lee committed:
address feedback

devin-ai-integration[bot] committed:
feat(bookings): add smart navigation for calendar view

- Add sort option to tRPC bookings.get schema and handler
- Create NAVIGATION_PROBE_WINDOW_MONTHS constant (3 months)
- Create useNearestFutureBooking hook to find nearest future booking
- Create useNearestPastBooking hook to find nearest past booking
- Update useCalendarNavigationCapabilities to use probe results
- Update BookingCalendarContainer to wire up probe hooks

This enables the booking details sheet to:
- Disable next/prev buttons when no bookings exist in that direction
- Jump directly to the week containing the nearest booking

Co-Authored-By: eunjae@cal.com <hey@eunjae.dev>

eunjae-lee committed:
fix type error

eunjae-lee committed:
Merge branch 'main' into fix/bookings-redesign-2

sean-brydon committed:
feat: add booking selection state when using slideover (#25637)

* implement booking selection when using slideover

* remove pixel shift

* fix

* auto scroll to selected event on calendar

* make DateValues header sticky when scrolling

---------

Co-authored-by: Eunjae Lee <hey@eunjae.dev>

eunjae-lee committed:
prefetch previous / next weeks on calendar view

hbjORbj reviewed (dismissed):
left a question!

sean-brydon reviewed (dismissed):
LGTM all feedback has been addressed - would be nice if we can get another approval before merging - touching a pretty big part of the app

hbjORbj reviewed (dismissed):
code nits

eunjae-lee commented on apps/web/modules/bookings/components/BookingListContainer.tsx:
```diff
@@ -180,7 +187,7 @@ function BookingListInner({
 
         <DataTableSegment.Select shortLabel />
-        {bookingsV3Enabled && <ViewToggleButton />}
+        {bookingsV3Enabled && <ViewToggleButton bookingsV3Enabled={bookingsV3Enabled} />}
```
yeah agree it's a bit weird. the thing is we need `bookingsV3Enabled` variable within the component to handle some stuff (a bit of prop drilling here).

On this particular component, we want to hide ViewToggleButton if bookingsV3Enabled === false.

<img width="1114" height="881" alt="Screenshot 2025-12-15 at 15 42 10" src="https://github.com/user-attachments/assets/3a16a82d-91f0-4df2-a02a-a860a507f87c" />

eunjae-lee committed:
clean up classes

eunjae-lee committed:
handle "fetched & but no data" situation more correctly in useCalendarAutoSelector

eunjae-lee committed:
Merge branch 'main' into fix/bookings-redesign-2

hbjORbj reviewed (approved):
lets go
